mesa/visualization/components/matplotlib.py

```diff
@@ -146,94 +186,180 @@ def draw_property_layers(ax, space, propertylayer_portrayal, model):
             )
 
 
-def _draw_grid(space, space_ax, agent_portrayal, propertylayer_portrayal, model):
-    if propertylayer_portrayal:
-        draw_property_layers(space_ax, space, propertylayer_portrayal, model)
+def collect_agent_data(
+    space: OrthogonalGrid | HexGrid | Network | ContinuousSpace,
+    agent_portrayal: Callable,
+):
+    """Collect the plotting data for all agents in the space.
+
+    Args:
+        space: The space containing the Agents.
+        agent_portrayal: A callable that is called with the agent and returns a dict
+        loc: a boolean indicating whether to gather agent x, y data or not
+
+    Notes:
+        agent portray dict is limited to s (size of marker), c (color of marker, and marker (marker style)
+        see `Matplotlib`_.
+
+
+    .. _Matplotlib: https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html
+
+    """
+    default_size = (180 / max(space.width, space.height)) ** 2
+
+    arguments = {"x": [], "y": [], "s": [], "c": [], "marker": []}
+    for agent in space.agents:
+        portray = agent_portrayal(agent)
+        loc = agent.pos
+        if loc is None:
+            loc = agent.cell.coordinate
+        x, y = loc
+        arguments["x"].append(x)
+        arguments["y"].append(y)
+
+        arguments["s"].append(portray.get("s", default_size))
+        arguments["c"].append(portray.get("c", "tab:blue"))
+        arguments["marker"].append(portray.get("marker", "o"))
+
+    return {k: np.asarray(v) for k, v in arguments.items()}
+
+
+def draw_orthogonal_grid(
+    space: OrthogonalGrid,
+    agent_portrayal: Callable,
+    propertylayer_portrayal: Callable,
+    model,
+):
+    """Visualize a orthogonal grid..
+
+    Args:
+        space: the space to visualize
+        agent_portrayal: a callable that is called with the agent and returns a dict
+        propertylayer_portrayal: a callable that is called with the agent and returns a dict
+        model: a model instance
+
+    Returns:
+        A Figure and Axes instance
 
-    agent_data = _get_agent_data(space, agent_portrayal)
+    """
+    arguments = collect_agent_data(space, agent_portrayal)
+
+    fig = Figure()
+    ax = fig.add_subplot(111)
 
-    space_ax.set_xlim(0, space.width)
-    space_ax.set_ylim(0, space.height)
-    _split_and_scatter(agent_data, space_ax)
+    ax.set_xlim(-0.5, space.width - 0.5)
+    ax.set_ylim(-0.5, space.height - 0.5)
 
     # Draw grid lines
-    for x in range(space.width + 1):
-        space_ax.axvline(x, color="gray", linestyle=":")
-    for y in range(space.height + 1):
-        space_ax.axhline(y, color="gray", linestyle=":")
+    for x in np.arange(-0.5, space.width - 0.5, 1):
+        ax.axvline(x, color="gray", linestyle=":")
+    for y in np.arange(-0.5, space.height - 0.5, 1):
+        ax.axhline(y, color="gray", linestyle=":")
```

Comment:
Not necessarily this PR, but I think we should add a boolean `draw_gridlines` or something like that to enable/disable this behavior.

Could also be `gridline_dict`, which if empty doesn't draw the gridlines, and otherwise takes a color, linestyle and width for example.

Reply:
As also indicated in #2389, yes, we need to find a clean solution for both view limits, grid lines, and possible user-specific post plotting operations. I'll leave that out of this PR.